source/common/network/listener_filter_buffer_impl.cc

```diff
@@ -0,0 +1,99 @@
+#include "source/common/network/listener_filter_buffer_impl.h"
+
+#include <string>
+
+namespace Envoy {
+namespace Network {
+
+ListenerFilterBufferImpl::ListenerFilterBufferImpl(IoHandle& io_handle,
+                                                   Event::Dispatcher& dispatcher,
+                                                   ListenerFilterBufferOnCloseCb close_cb,
+                                                   ListenerFilterBufferOnDataCb on_data_cb,
+                                                   uint64_t buffer_size)
+    : io_handle_(io_handle), dispatcher_(dispatcher), on_close_cb_(close_cb),
+      on_data_cb_(on_data_cb), buffer_(std::make_unique<uint8_t[]>(buffer_size)),
+      base_(buffer_.get()), buffer_size_(buffer_size) {
+  // If the buffer_size not greater than 0, it means that doesn't expect any data.
+  ASSERT(buffer_size > 0);
+
+  io_handle_.initializeFileEvent(
+      dispatcher_, [this](uint32_t events) { onFileEvent(events); },
+      Event::PlatformDefaultTriggerType, Event::FileReadyType::Read);
+}
+
+const Buffer::ConstRawSlice ListenerFilterBufferImpl::rawSlice() const {
+  Buffer::ConstRawSlice slice;
+  slice.mem_ = base_;
+  slice.len_ = data_size_;
+  return slice;
+}
+
+void ListenerFilterBufferImpl::drain(uint64_t length) {
+  if (length == 0) {
+    return;
+  }
+
+  ASSERT(length <= data_size_);
+
+  uint64_t read_size = 0;
+  while (1) {
+    auto result = io_handle_.recv(base_, length - read_size, 0);
+    ENVOY_LOG(trace, "recv returned: {}", result.return_value_);
+
+    // The socket buffer is expected to have the data. so the
+    // recv doesn't expected to fail.
+    ASSERT(result.ok());
```

Comment:
Are there any error conditions on a socket (client sends RST maybe? Some administrative operation on the host?) that could make the operation fail? It may be safer to break out of the loop if this fails, even though it would be unexpected, so avoid an infinite loop.

Reply:
agree that it is safer to break out of the loop. Do we want to throw an exception here? or we want to close the connection? sounds like second on is better.

Reply:
Close the connection. Exceptions are only allowed while handling configuration.